### PR TITLE
536 follow up merton

### PR DESCRIPTION
**Pull request by jacobvjk**
closes ADO 536 https://dev.azure.com/2DegreesInvesting/2DegreesInvesting/_workitems/edit/536

follow up on https://github.com/2DegreesInvesting/StressTestingModelDev/pull/307

Apologies in advance for yet another PR that has grown larger than expected. I should have made a separate one for the graphs.. @maurolepore quite a bit of the changes only became clear while developing, since this is still experimental.

This PR:
- calculates the overall PD changes for maturities of 1 to 5 years based on the entire NPV of the starting value equities `R/calculate_overall_pd_changes.R`
- uses the overall pd changes in the expected loss calculations, grouping all maturities greater than 5 years in the 5 year bucket
- calculates separate annual PD changes for all years of the analysis in a cumulative way, using start year until year t for every calculation to derive PD change pathways `R/calculate_annual_pd_changes.R`
- go back to defining PD changes in absolute terms, not relative terms
- extracts some data preparation steps from merton model into lower level helper functions `R/create_empty_result_df_pd_changes.R` and `R/add_cols_result_df_pd_changes.R` so the main logic is less buried.
- removes the file `R/merton.R` in favor of the two separate ones above
- writes the outcomes of the credit risk calculation to the result directory of the project (see `stress_test_model_loan_book.R`)
- corrects the future profits used in determining the input equity values by filtering out profits pre shock. This ensures the shocks magnitude changes in expectation with the shock year.
- adds graphing functions for annual and overall credit risk `R/annual_pd_change_company_technology.R`, `R/annual_pd_change_technology_shock_year.R`, `R/overall_pd_change_company_technology.R`, `R/overall_pd_change_technology_shock_year.R`
- imports new `ggplot2` functions and declares `r2dii.plot` a dependency highlighting its GH remote
- adds tests for all new functions

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -33,10 +33,14 @@ Imports:
...
     purrr,
+    r2dii.plot,
     readr,
     rlang,
-    testthat,
+    stringr,
     tidyr,
     withr,
     zoo
+Remotes: 2DegreesInvesting/r2dii.plot
```
Comment (maurolepore) [inline DESCRIPTION:43]: Cool, just be aware that we are changing the API radically until the first release on CRAN. You may need some way to know if the interface changes in a way that breaks your code (i.e. an integration test -- like an .Rmd file that calls the r2dii.plot functions you want to use and breaks on CI if it fails to render.).

File: R/add_cols_result_df_pd_changes.R [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#' @param data A dataframe that contains the result data from the baseline
+#'   Merton model
+#' @param horizon A cahracter vector of length 1 indicating whether to return
+#'   a result obejct for overall or yearly results. Must be "annual" or "overall"
+#'
+#' @family helper functions
+#'
+#' @export
+
+add_cols_result_df_pd_changes <- function(data,
```
Comment (maurolepore) [inline R/add_cols_result_df_pd_changes.R:13]: Style?

File: R/add_cols_result_df_pd_changes.R [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                                          horizon = NULL) {
+
+  force(data)
+  horizon %||% stop("Must provide input for 'horizon'", call. = FALSE)
+
+  horizon_allowed <- horizon %in% c(
+    "annual",
+    "overall"
+  )
+  stopifnot(horizon_allowed)
```
Comment (maurolepore) [inline R/add_cols_result_df_pd_changes.R:23]: The vertical alignment is crates nice git diffs but might be an overkill here and separates the definition and use of `horizon_allowed` by 3 lines.

File: R/add_cols_result_df_pd_changes.R [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  if (horizon == "overall") {
```
Comment (maurolepore) [inline R/add_cols_result_df_pd_changes.R:25]: Remember if takes an atomic input but `==` can return a longer vector. You may check the length of `horizon` before `if` or use `identical()` inside `if`.

File: R/add_cols_result_df_pd_changes.R [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    )
+  } else {
+    data_has_expected_columns <- all(
+      c(
+        "investor_name", "portfolio_name", "scenario_name", "scenario_geography",
+        "id", "company_name", "ald_sector", "technology", "year",
+        "equity_t_baseline", "equity_t_late_sudden", "debt", "volatility",
+        "risk_free_rate", "term", "Maturity", "Vt", "St", "Dt", "Survival"
+      ) %in% colnames(data)
+    )
```
Comment (maurolepore) [inline R/add_cols_result_df_pd_changes.R:42]: You may DRY this duplication and get a clearer overview of what's the difference between these two.

Maybe

```r
common <- c("a", "b")
expected <- c(common, "d")
if (<something happens>) {
  expected <- c(common, "f")
}
expected

File: R/annual_pd_change_company_technology.R [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  stopifnot(data_has_expected_columns)
+
+  shock_year <- stringr::str_c("Carbon balance ", shock_year)
+
+  plot <- data %>%
+    dplyr::filter(
+      .data$company_name %in% .env$company_filter &
+        .data$scenario_name == .env$shock_year &
+        .data$scenario_geography %in% geography_filter
+    ) %>%
```
Comment (maurolepore) [inline R/annual_pd_change_company_technology.R:39]: Style: With `filter()` it's rare to use `&` and more common to use `, `.  With `subset()` you are forced to use `&`.

**Review by maurolepore — APPROVED** [review]

**Review by maurolepore — APPROVED** [review]
It's quite bigger than I can review right now but still I made some comments that might help a bit.

File: DESCRIPTION (hunk 1 above)
Comment (jacobvjk) [reply DESCRIPTION:43]: I think it's fine for no, it's only being used in non-essential QA graphs for now. If they break, this won't impact any of the analysis. But I will keep my eyes open.

File: R/add_cols_result_df_pd_changes.R (hunk 2 above)
Comment (jacobvjk) [reply R/add_cols_result_df_pd_changes.R:13]: wasn't the data argument supposed to be on the same line?

File: R/add_cols_result_df_pd_changes.R (hunk 5 above)
Comment (jacobvjk) [reply R/add_cols_result_df_pd_changes.R:42]: This is a nice idea!

**Pull request merged**